qutebrowser/browser/downloads.py

```diff
@@ -618,314 +546,89 @@ def set_filename(self, filename):
-            self.fileobj.write(self._reply.readAll())
-        if self.autoclose:
-            self.fileobj.close()
-        self.successful = self._reply.error() == QNetworkReply.NoError
-        self._reply.close()
-        self._reply.deleteLater()
-        self._reply = None
-        self.finished.emit()
-        self.done = True
-        log.downloads.debug("Download {} finished".format(self.basename))
-        self.data_changed.emit()
-
-    @pyqtSlot()
-    def _on_reply_finished(self):
-        """Clean up when the download was finished.
-
-        Note when this gets called, only the QNetworkReply has finished. This
-        doesn't mean the download (i.e. writing data to the disk) is finished
-        as well. Therefore, we can't close() the QNetworkReply in here yet.
+            cmdline: Passed to DownloadItem.open_file().
         """
-        if self._reply is None:
-            return
-        self._read_timer.stop()
-        self.stats.finish()
-        is_redirected = self._handle_redirect()
-        if is_redirected:
+        if not self.successful:
+            log.downloads.debug("{} finished but not successful, not opening!"
+                                .format(self))
             return
-        log.downloads.debug("Reply finished, fileobj {}".format(self.fileobj))
-        if self.fileobj is not None:
-            # We can do a "delayed" write immediately to empty the buffer and
-            # clean up.
-            self._finish_download()
+        self.open_file(cmdline)
 
-    @pyqtSlot()
-    def _on_ready_read(self):
-        """Read available data and save file when ready to read."""
-        if self.fileobj is None or self._reply is None:
-            # No filename has been set yet (so we don't empty the buffer) or we
-            # got a readyRead after the reply was finished (which happens on
-            # qute:log for example).
-            return
-        if not self._reply.isOpen():
-            raise OSError("Reply is closed!")
-        try:
-            self.fileobj.write(self._reply.readAll())
-        except OSError as e:
-            self._die(e.strerror)
-
-    @pyqtSlot('QNetworkReply::NetworkError')
-    def _on_reply_error(self, code):
-        """Handle QNetworkReply errors."""
-        if code == QNetworkReply.OperationCanceledError:
-            return
-        else:
-            self._die(self._reply.errorString())
+    def set_target(self, target):
+        """Set the target for a given download.
 
-    @pyqtSlot()
-    def _on_read_timer_timeout(self):
-        """Read some bytes from the QNetworkReply periodically."""
-        if not self._reply.isOpen():
-            raise OSError("Reply is closed!")
-        data = self._reply.read(1024)
-        if data is not None:
-            self._buffer.write(data)
+        Args:
+            target: The usertypes.DownloadTarget for this download.
+        """
+        if isinstance(target, usertypes.FileObjDownloadTarget):
```

Comment:
Since this is the abstract base class, wouldn't it be better to just call `self._set_fileobj` here and let the QWebEngineDownloadItem's `_set_fileobj` take care of raising the `UnsupportedOperationError`? Seems weird to have exactly this target ruled out in the base, when `FileDownloadTarget` and `OpenFileDownloadTarget` are handled fine here.